controller/deployments_urlprovider.go

```diff
@@ -91,130 +28,31 @@ var _ kubernetes.BaseURLProvider = (*tenantURLProvider)(nil)
 	if err != nil {
-		log.Error(ctx, map[string]interface{}{
-			"err": err,
-		}, "error accessing Auth server")
-		return nil, errs.Wrap(err, "error creating Auth client")
-	}
-	up.TokenRetriever = &tokenRetriever{
-		authClient: authClient,
-		context:    ctx,
+		return nil, err
 	}
 
-	// if we're not using a proxy then the API URL is actually the cluster of namespace 0,
-	// so the apiToken should be the token for that cluster.
-	// there should be no defaults, but that's deferred later
-	if len(proxyURL) == 0 {
-		tokenData, err := up.TokenForService(up.apiURL)
-		if err != nil {
-			return nil, err
-		}
-		up.apiToken = *tokenData
-	}
 	return up, nil
 }
 
-// newTenantURLProviderFromTenant create a provider from a UserService object
-func newTenantURLProviderFromTenant(t *app.UserService, token string, proxyURL string) (*tenantURLProvider, error) {
-
-	if t.ID == nil {
-		log.Error(nil, map[string]interface{}{}, "app.UserService is malformed: no ID field")
-		return nil, errs.New("app.UserService is malformed: no ID field")
-	}
-
-	if t.Attributes == nil {
-		log.Error(nil, map[string]interface{}{
-			"tenant": *t.ID,
-		}, "app.UserService is malformed: no Attribute field ID=%s", *t.ID)
-		return nil, errs.Errorf("app.UserService is malformed: no Attribute field (ID=%s)", *t.ID)
-	}
-
-	if len(t.Attributes.Namespaces) == 0 {
-		log.Error(nil, map[string]interface{}{
-			"tenant": *t.ID,
-		}, "this tenant has no namespaces: %s", *t.ID)
-		return nil, errs.Errorf("app.UserService is malformed: no Namespaces (ID=%s)", *t.ID)
-	}
-
-	defaultNamespace := t.Attributes.Namespaces[0]
-	namespaceMap := make(map[string]*app.NamespaceAttributes)
-	for i, namespace := range t.Attributes.Namespaces {
-		namespaceMap[*namespace.Name] = t.Attributes.Namespaces[i]
-		if namespace.Type != nil && *namespace.Type == "user" {
-			defaultNamespace = namespace
-		}
-	}
-
-	defaultClusterURL := *defaultNamespace.ClusterURL
-
-	if len(proxyURL) != 0 {
-		// all non-metric API calls go via the proxy
-		defaultClusterURL = proxyURL
-	}
-
+// NewProxyURLProvider create a provider from a UserService object (exposed for testing)
+func NewProxyURLProvider(token string, proxyURL string) (kubernetes.BaseURLProvider, error) {
 	provider := &tenantURLProvider{
-		apiURL:     defaultClusterURL,
-		apiToken:   token,
-		tenant:     t,
-		namespaces: namespaceMap,
+		apiURL:   proxyURL,
+		apiToken: token,
 	}
 	return provider, nil
 }
 
-// NewTenantURLProviderFromTenant create a provider from a UserService object (exposed for testing)
-func NewTenantURLProviderFromTenant(t *app.UserService, token string, proxyURL string) (kubernetes.BaseURLProvider, error) {
-	return newTenantURLProviderFromTenant(t, token, proxyURL)
-}
-
-// GetEnvironmentMapping returns a map whose keys are environment names, and values are the Kubernetes namespaces
-// that represent those environments
-func (up *tenantURLProvider) GetEnvironmentMapping() map[string]string {
```

Comment:
We will still need this method to translate environment names to Kubernetes namespaces using the information in Tenant.

Reply:
Unfortunately, I was hoping to get rid of Tenant in our code, but it seems it's not possible.
I think this API as is is unusable because of the issues @ebaron pointed out, so I propose withdrawing it.  The current code will work via the proxy if so configured; this was mainly  a cleanup PR.  A proper refactoring would leave most of this PR as I have it here (i.e. simplify the URLProvider code), but also add a TenantClient that provides the mapping we reuire from the Tenant service.